### PR TITLE
Encode filter parameter as a base64-encoded JSON string in List requests

### PR DESCRIPTION
**Pull request by neuromage**
Generate new swagger definitions and use these to generate the frontend
APIs using `npm run apis`.

This is to support filtering in List requests, as the current
grpc-gateway swagger generator tool does not support repeated fields in
requests used in GET endpoints.

Using POST lets us get around the problem in issue grpc-ecosystem/grpc-gateway#492

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/563)

### CODE REVIEW

**Comment by yebrahim** [comment]
The frontend needs to change the API calls to pass the object. I can help take a look at this soon.

File: frontend/src/apis/experiment/api.ts [hunk 1]
```diff
@@ -191,45 +357,6 @@ export interface ProtobufEmpty {
  */
 export const ExperimentServiceApiFetchParamCreator = function (configuration?: Configuration) {
     return {
-        /**
-         * 
-         * @param {ApiExperiment} body 
-         * @param {*} [options] Override http request option.
-         * @throws {RequiredError}
-         */
-        createExperiment(body: ApiExperiment, options: any = {}): FetchArgs {
```
Comment (yebrahim) [inline frontend/src/apis/experiment/api.ts]: Why are these removed?

Comment (neuromage) [reply frontend/src/apis/experiment/api.ts]: Thanks for the heads up. This was indeed messed up, because both create and list used the same endpoints with differing verbs (POST and GET).

I decided to go another route: encode the filter as a simple JSON string and put it in a single parameter. Maybe this was what you were alluding to earlier, and I think this should work better, as it doesn't break any of the existing semantics. See what you think.

File: backend/src/apiserver/server/experiment_server.go [hunk 2]
```diff
@@ -44,7 +46,23 @@ func (s *ExperimentServer) ListExperiment(ctx context.Context, request *api.List
 	if request.PageToken != "" {
 		opts, err = list.NewOptionsFromToken(request.PageToken, int(request.PageSize))
 	} else {
-		opts, err = list.NewOptions(&model.Experiment{}, int(request.PageSize), request.SortBy, request.Filter)
+		var f *api.Filter
```
Comment (yebrahim) [inline backend/src/apiserver/server/experiment_server.go]: Any chance we can reuse this code instead of copying it?

Comment (neuromage) [reply backend/src/apiserver/server/experiment_server.go]: Thanks! I consolidated this logic into a new function, and also added tests for it. PTAL.

File: backend/src/apiserver/server/list_request_util_test.go [hunk 3]
```diff
@@ -175,3 +177,40 @@ func TestParseSortByQueryString_StringTooLong(t *testing.T) {
...
+		Predicates: []*api.Predicate{
+			&api.Predicate{
+				Key: "testkey", Op: api.Predicate_EQUALS,
+				Value: &api.Predicate_StringValue{StringValue: "testvalue"},
+			},
+		},
+	}
+
+	got, err := parseAPIFilter(in)
+	if !cmp.Equal(got, want) || err != nil {
```
[the remark this answers is not in the document]
Comment (neuromage) [reply backend/src/apiserver/server/list_request_util_test.go]: True, but I actually prefer the cmp package, as it prints a more precise diff. It's also maintained and recommended by the Go team.

File: backend/api/go_http_client/experiment_client/experiment_service/list_experiment_parameters.go [hunk 4]
```diff
@@ -126,6 +132,17 @@ func (o *ListExperimentParams) SetHTTPClient(client *http.Client) {
...
 
+// WithFilter adds the filter to the list experiment params
+func (o *ListExperimentParams) WithFilter(filter *string) *ListExperimentParams {
+	o.SetFilter(filter)
+	return o
+}
+
+// SetFilter adds the filter to the list experiment params
+func (o *ListExperimentParams) SetFilter(filter *string) {
+	o.Filter = filter
```
Comment (rileyjbauer) [inline backend/api/go_http_client/experiment_client/experiment_service/list_experiment_parameters.go]: If this function is so simple, and not widely used, why not just inline it?

Comment (neuromage) [reply backend/api/go_http_client/experiment_client/experiment_service/list_experiment_parameters.go]: This is auto-generated by the swagger tool. Everything under api/go_http_client and api/go_client is autogenerated.

Comment (rileyjbauer) [reply backend/api/go_http_client/experiment_client/experiment_service/list_experiment_parameters.go]: Ah, sorry, I didn't realize they were auto-generated

Comment (neuromage) [reply backend/api/go_http_client/experiment_client/experiment_service/list_experiment_parameters.go]: Sorry, yeah, it's definitely not obvious :-)

File: backend/src/apiserver/server/experiment_server.go [hunk 5]
```diff
@@ -44,7 +44,12 @@ func (s *ExperimentServer) ListExperiment(ctx context.Context, request *api.List
 	if request.PageToken != "" {
 		opts, err = list.NewOptionsFromToken(request.PageToken, int(request.PageSize))
 	} else {
-		opts, err = list.NewOptions(&model.Experiment{}, int(request.PageSize), request.SortBy, request.Filter)
+		f, err := parseAPIFilter(request.Filter)
```
Comment (IronPan) [inline backend/src/apiserver/server/experiment_server.go]: not quite understand the if else statement here. does it means only one of pagetoken and filter can exist?
Are filtered results paginated?

File: backend/src/apiserver/server/list_request_util.go [hunk 6]
```diff
@@ -157,3 +158,27 @@ func deserializePageToken(pageToken string) (*common.Token, error) {
...
+
+// parseAPIFilter attempts to decode a base-64 encoded JSON-stringified api
+// filter object. An empty string is considered valid input, and equivalent to
+// the nil filter, which trivially does nothing.
+func parseAPIFilter(encoded string) (*api.Filter, error) {
+	if encoded == "" {
+		return nil, nil
+	}
+
+	errorF := func(err error) (*api.Filter, error) {
```
Comment (IronPan) [inline backend/src/apiserver/server/list_request_util.go]: is any good reason of having this function instead of returning errors directly?

File: backend/src/apiserver/server/list_request_util.go [hunk 7]
```diff
@@ -157,3 +158,27 @@ func deserializePageToken(pageToken string) (*common.Token, error) {
...
+	b, err := base64.URLEncoding.DecodeString(encoded)
+	if err != nil {
+		return errorF(err)
+	}
+
+	f := &api.Filter{}
+	if err := jsonpb.UnmarshalString(string(b), f); err != nil {
+		return errorF(err)
+	}
+	return f, nil
```
Comment (IronPan) [inline backend/src/apiserver/server/list_request_util.go]: can we add additional validation to make sure the filter format is correct?
e.g. value missing, op missing etc.

File: backend/src/apiserver/server/experiment_server.go (hunk 5 above)
Comment (neuromage) [reply backend/src/apiserver/server/experiment_server.go]: pagetoken encapsulates both sorting and filtering, so if you already have a non-empty page token, we ignore further filtering + sorting criteria.

File: backend/src/apiserver/server/list_request_util.go (hunk 6 above)
Comment (neuromage) [reply backend/src/apiserver/server/list_request_util.go]: Avoids duplicate in error handling, since the message is the same. It's a very idiomatic Go thing to do.

File: backend/src/apiserver/server/list_request_util.go (hunk 7 above)
Comment (neuromage) [reply backend/src/apiserver/server/list_request_util.go]: The api filter will get parsed into a filter.Filter object that can actually be used to do filtering. That parsing will sanity check to ensure the filter is well formed and return an error otherwise.

**Comment by neuromage** [comment]
Thanks @rileyjbauer for pointing out that the Filter definitions for Swagger went missing with this change. This was caused by the fact that Filter is only implicitly used by the API, so protoc-gen-swagger will not generate Swagger files correctly for this proto. I modified the Makefile to hack in a dummy service during generation to ensure we get the Swagger definitions. And I updated the `npm run apis` script to generate TS interface definitions for filter as well. 

PTAL.

File: backend/api/Makefile [hunk 8]
```diff
@@ -52,6 +52,13 @@ all: dependencies
 		--grpc-gateway_out=logtostderr=true:go_client \
 		*.proto
 
+  # Filter.proto is implicitly used by clients and server, and transmitted as a
```
[the remark this answers is not in the document]
Comment (neuromage) [reply backend/api/Makefile]: Done.

**Comment by neuromage** [comment]
Yeah, it's a pretty ugly hack. 

I did have the dummy service in initially, but it was generating a corresponding client/server code in Go, which didn't seem like the right thing to do. Hence, this hack :-(

**Comment by yebrahim** [comment]
How about setting up "go vet -shadow" for catching these errors. I think shadowing is bad practice anyway. 
@neuromage @IronPan

**Comment by neuromage** [comment]
Agreed! We can set this up in Travis as part of the presubmit. If you're ok with it, I can try to do that, and also fix any existing shadow errors.

**Comment by neuromage** [comment]
yes, definitely, I meant in a separate PR. This is one is good to go @yebrahim @IronPan

File: backend/api/Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ark-kun) [reply backend/api/Makefile]: > nit: indentation (this file uses tabs for some reason).

It's not a nit. Makefile requires tabs for receipt prefixes and will misbehave if they are not there.

[the remark this answers is not in the document]
Comment (yebrahim) [reply backend/api/Makefile]: Weird, but good to know.

**Comment by IronPan** [comment]
/lgtm
/approve
assuming that you will have a separate PR to validate the query string as we discussed offline
- If page token doesn't exist, validate the filter and sort by fields.
- If page token exist, either
   - filter and sort by fields exist and they are valid and equals to the one encoded in page token
   - or they can be opt-out and no validation is needed in this case.
